README.md
## 项目简介
微信支付服务,支持微信中H5-jsapi支付,微信小程序支付

## 文档
[境内普通商户-微信小程序支付](https://pay.weixin.qq.com/wiki/doc/api/wxa/wxa_api.php?chapter=7_4&index=3)
[微信小程序支付中用到的微信小程序登录](https://developers.weixin.qq.com/miniprogram/dev/framework/open-ability/login.html)

## 打包发布流程
1. 编译及打包发布
```
npm install
npm run build
npm run tar
scp bz2-server.tar.bz2 root@qingshansi:/data/nodejs/jtb-api-server/
```
2. 服务器部署
```
cd /data/nodejs/jtb-api-server/
tar -jxvf bz2-server.tar.bz2
npm install

# 启动方式1: node方式
node sdist

# 启动方式2: pm2方式
DEBUG="app:*" pm2 start sdist --name jtb-api-server
pm2 save
pm2 startup
```

## docker打包及运行
1. nodejs服务打包
docker build . -t windsome/wechat-payment:1.0.1

2. 用docker运行
docker run -p 33310:3310 windsome/wechat-payment:1.0.1

## k8s部署
1. 从目录创建ConfigMap,见<https://www.jianshu.com/p/cf8705a93c6b>,需要k8s1.10.0以上版本才支持二进制文件.
```
$ scp k8s/conf.single root@xiaobei1:/data/k8s/ # 将配置文件目录拷贝的k8s服务器上.
$ kubectl create configmap wepay-cm --from-file=/data/k8s/conf.single/
configmap "wepay-cm" created
$ kubectl get configmap wepay-cm -o go-template='{{.data}}'
map[config.json:config.json
 apiclient_cert_1415522602.p12:apiclient_cert_1415522602.p12
 apiclient_cert_1525363461.p12:apiclient_cert_1525363461.p12
]
```
2. 使用Volume将ConfigMap作为文件或目录挂载
将创建的ConfigMap直接挂载至Pod的/data/wechat-payment/config目录下，其中每一个key-value键值对都会生成一个文件，key为文件名，value为内容。
```
---
apiVersion: v1
kind: Service
metadata:
  labels:
    k8s-app: wechat-payment
  name: wechat-payment
  namespace: default
spec:
  ports:
  - name: wechat-payment-port-80
    port: 80
    targetPort: 3310
    protocol: TCP
  selector:
    k8s-app: wechat-payment
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    k8s-app: wechat-payment
  name: wechat-payment
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      k8s-app: wechat-payment
  template:
    metadata:
      labels:
        k8s-app: wechat-payment
      name: wechat-payment
    spec:
      containers:
      - image: windsome/wechat-payment:latest
        name: wechat-payment
        imagePullPolicy: Always
        volumeMounts:
        - name: config-volume
          mountPath: /data/wechat-payment/config
      volumes:
      - name: config-volume
        configMap:
          name: wepay-cm
```

## 配置文件构造
配置文件的内容:
```
{
  "name":"此支付的名称,方便分辨,没有实际意义",
  "env": "<H5>or<WXAPP>or<APP>",
  "type": "<普通商户>or<服务商>",
  "appId": "<your_appId>",
  "mchId": "<your_mchId>",
  "notifyUrl": "<接收微信支付的通知的URL,本服务需要暴露给外界,并需要在微信支付后台设置>",
  "partnerKey": "<your_partnerKey>",
  "pfx": "<微信的P12文件>",
  "passphrase": "<your_passphrase>一般与mchId相同",
  "_payUrl": "<支付目录,在此SDK中不用,在H5支付中需配置在微信支付后台>",
  "cbNotifyUrl": "<本服务接受微信通知后,再通知其他微服务处理实际订单的URL>"
}
```
单个支付:
```
{
  env: 'H5',
  type: '普通商户',
  appId: 'wx9af096e5f111831a',
  mchId: '1525312341',
  notifyUrl: 'http://m.qingshansi.cn/apis/v1/wepay/pay_notify',
  partnerKey: '7gwY3krPn33c1lBzO11YXA1TdbSsjXT1',
  pfx: fs.readFileSync(__dirname + '/apiclient_cert_1525312341.p12'),
  passphrase: '1525312341',
  _payUrl: 'http://m.qingshansi.cn/eshop',
  cbNotifyUrl: "http://sharks-api/apis/v1/mqtt/pay_notify"
}
```
多个支付:
```
[{
  env: 'WXAPP',
  type: '普通商户',
  appId: 'wx9af096e5f111831a',
  mchId: '1525312341',
  notifyUrl: 'http://m.qingshansi.cn/apis/v1/wepay/pay_notify',
  partnerKey: '7gwY3krPn33c1lBzO11YXA1TdbSsjXT1',
  pfx: fs.readFileSync(__dirname + '/apiclient_cert_1525312341.p12'),
  passphrase: '1525312341',
  _payUrl: 'http://m.qingshansi.cn/eshop',
  cbNotifyUrl: "http://sharks-api/apis/v1/mqtt/pay_notify"
},{
  env: 'APP',
  type: '普通商户',
  appId: 'wx9af096e5f111831a',
  mchId: '1525312341',
  notifyUrl: 'http://m.qingshansi.cn/apis/v1/wepay/pay_notify',
  partnerKey: '7gwY3krPn33c1lBzO11YXA1TdbSsjXT1',
  pfx: fs.readFileSync(__dirname + '/apiclient_cert_1525312341.p12'),
  passphrase: '1525312341',
  _payUrl: 'http://m.qingshansi.cn/eshop',
  cbNotifyUrl: "http://sharks-api/apis/v1/mqtt/pay_notify"
}]
```

## 用apidoc生成接口文档
```
npm run apidoc
```

## 供微信通知用接口
是`POST /apis/v1/wepay/pay_notify`,其中会调用其他服务的订单处理服务.此订单处理服务在`cbNotifyUrl`中指定.


## API接口 (使用HTTP)
本考虑使用thrift,后来考虑到主要性能瓶颈在于调用微信服务器,而不在于内部通信,就算用thrift加快内部通信速度,于实际总调用时长无太大关系.
1. `POST /apis/v1/wepay/get_pay_request_params`,调用统一下单接口,返回供微信H5使用的支付
```
参数:
{
  id: '5ba27cc3a70db45dd108b53f', // 订单id,
  fee: 1200, // 订单金额,以分为单位
  body: '商品描述:袜子1件+衣服2件', // 必选, string(128) 商品描述
  detail: '商品详情:2019赛季球衣 蓝色 L 高领 2件; 球袜 红色 小 1件' // 可选, string(600) 商品详情
  attach: '附加数据:福州路店', // 可选, string(127) 附加数据
}
正确返回:
{
  errcode: 0,
  result: {
    xxx
  }
}
错误返回:
{
  errcode: !=0
}
```
2. `POST /apis/v1/wepay/refund`,调用退款接口, 
```
参数:
{
  out_refund_no: '5ba27cc3a112233554444331', //退款单ID,在系统中一般有张退款单表
  out_trade_no: '5ba27cc3a70db45dd108b53f', // 订单id,
  total_fee: 1200, // 订单总金额,
  refund_fee: 1100, // 本次退款金额,以分为单位,必须小于等于total_fee,并且本订单所有退款总额不超过total_fee
}
正确返回:
{
  errcode: 0,
  result: {
    xxx
  }
}
错误返回:
{
  errcode: !=0
}
```
3. `POST /apis/v1/wepay/get_sign_key` 获取测试用sign_key
```
参数:无
正确返回:
{
  errcode: 0,
  result: {
    xxx
  }
}
错误返回:
{
  errcode: !=0
}
```

## TODO
1. 微信小程序支付
2. 微信APP支付


CHANGELOG.md
Changelog
=========

1.0.1
-------------
### Features
* 增加cbNotifyUrl支持
* 
### Improvements
* 完善README.md中接口描述

1.0.0
-------------
### Features
* 初始化
* 可通过configmap给docker容器设置参数
* 
### Improvements
* 无


src/apis/apiWepay.js
import _debug from 'debug';
const debug = _debug('app:apiWepay');
import _ from 'lodash';
import getRawBody from 'raw-body';
import fs from 'fs';
import WechatPay from '../lib/wepay';
import Errcode, { EC } from '../Errcode';
import CommonRouterFix from './commonRouter';
import type from '../utils/type';
import { requestPost } from '../utils/_request';

const base_cfg_folder = '/data/wechat-payment/config/';
/**
 * 支付整体过程:
 * 1. 用户在客户端选中一个或一系列商品,生成一个订单.
 * 2. 给用户显示订单详情,根据订单状态,调用微信支付弹出支付确认框.
 * 3. 用户输入密码确认付款,客户端则处于支付结果等待页面(每隔1秒向服务器请求订单状态).
 * 4. 服务器端会收到微信发来的支付结果(微信主动调用http://m.qingshansi.cn/apis/v1/wepay/pay_notify)
 * 5. 服务器端处理微信服务器发来的支付结果,并处理,修改订单状态,返回微信处理结果.
 * 6. 客户端确认订单新状态后,提示客户成功还是失败.
 * 7. 订单成功后,如果是直接使用的,如VIP会员卡,则直接使用了.
 */
export default class Apis {
  constructor(app) {
    this.app = app;

    this.init();
    this.registerServices();
  }

  init = () => {
    debug('init WechatPay');
    try {
      let config = JSON.parse(fs.readFileSync( base_cfg_folder+'config.json'));
      let cfg0 = null;
      if (config) {
        let typeconfig = type(config);
        if (typeconfig === 'array') cfg0 = config[0];
        else cfg0 = config;
      }
      this.cfg = cfg0;
    } catch (error) {
      debug('error!', error);
    }
    if(this.cfg) {
      let pfx = fs.readFileSync(base_cfg_folder+this.cfg.pfx);
      this.cfg.pfx = pfx;
    }
    debug('wechat payment config: ', this.cfg);
    this.wepay = new WechatPay(this.cfg);
  };

  registerServices() {
    // init router for apis.
    let prefix = '/apis/v1/wepay';
    let router = require('koa-router')({ prefix });
    router.all('/pay_notify', this.payNotify);
    router.post('/get_pay_request_params', this.getPayRequestParams);
    router.post('/refund', this.refund);
    //sandbox
    router.post('/get_sign_key', this.getSandboxSignKey);

    CommonRouterFix(this.app, router, prefix);
  }

  ///////////////////////////////////////////////////
  // wechat pay.
  ///////////////////////////////////////////////////
  payNotify = async (ctx, next) => {
    let rawText = await getRawBody(ctx.req, {
      encoding: 'utf-8'
    });
    //debug ("get notify: ", rawText);
    try {
      let retobj = await this.wepay.notifyParse(rawText);
      debug('payNotify parsed:', retobj);
      if (!retobj) {
        throw new Error('pay notify get null!');
      }
      if (retobj.return_code != 'SUCCESS') {
        throw new Error(
          'error! pay notify error! return_code=' +
            retobj.return_code +
            ', return_msg=' +
            retobj.return_msg
        );
      }

      // notify others.
      let result = await this._notifyPayment(retobj);
      debug('_notifyPayment:', result);

      let xml = this.wepay.notifyResult({
        return_code: 'SUCCESS',
        return_msg: 'OK'
      });
      debug('payNotify process ok: ', xml);
      ctx.body = xml;
    } catch (e) {
      debug('payNotify error: ', e);
      let xml = this.wepay.notifyResult({
        return_code: 'FAIL',
        return_msg: 'FAIL'
      });
      ctx.body = xml;
    }
  };

  _notifyPayment = async (notification) => {
    let cbUrl = this.cfg && this.cfg.cbNotifyUrl;
    if (!cbUrl) {
      throw new Error('no cbNotifyUrl!');
    }
    let retobj = await requestPost(cbUrl, notification);
    if (!retobj) {
      throw new Error('no result from cbUrl!');
    }
    if (retobj.errcode) {
      throw retobj;
    }
    return retobj;
  }

  /**
   * @api {POST} /apis/v1/wepay/getPayRequestParams 获取订单wepay支付参数
   * @apiDescription `Content-Type="application/json"` 获取订单wepay支付参数
   * 1. 查询某id的订单,判断此订单是否为可支付状态.
   * 2. 构建微信订单信息.
   * 3. 调用wepay.getBrandWCPayRequestParams()获得支付参数.
   * 6. 返回结果
   * @apiName other-getPayRequestParams
   * @apiGroup apiWepay
   * @apiVersion 1.0.0
   * @apiContentType application/json
   * @apiParamExample {json} Request-Example:
   * {
   *  id: '5ba27cc3a70db45dd108b53f', // 订单id
   *  fee: 1200, // 订单金额,以分为单位
   *  body: '商品描述:袜子1件+衣服2件', // 必选, string(128) 商品描述
   *  detail: '商品详情:2019赛季球衣 蓝色 L 高领 2件; 球袜 红色 小 1件' // 可选, string(600) 商品详情
   *  attach: '附加数据:福州路店', // 可选, string(127) 附加数据
   * }
   */
  getPayRequestParams = async (ctx, next) => {
    let orderId = ctx.params.id || ctx.request.body.id || ctx.request.query.id;
    if (!orderId) {
      throw new Errcode('error! param order.id=null!', EC.ERR_PARAM_ERROR);
    }

    let args = ctx.request.body || {};
    let { fee, openid, body = '', detail = '', attach = '', ip } = args;
    ip = ip || ctx.request.ip;

    // 组装订单信息.
    if (!fee) {
      // 无需资金支付,直接修改状态为已支付.
      throw new Errcode(
        'error! no need pay, order.id=' + orderId,
        EC.ERR_ORDER_NO_NEED_PAID
      );
    }

    let trade = {
      body: body.substr(0, 127),
      detail: detail.substr(0, 600),
      attach: attach.substr(0, 127),
      out_trade_no: orderId,
      total_fee: fee,
      // sub_openid: openid,
      trade_type: 'JSAPI',
      spbill_create_ip: ip
    };
    if (this.cfg.type === '普通商户') trade = { ...trade, openid: openid };
    else trade = { ...trade, sub_openid: openid };

    debug('about to getBrandWCPayRequestParams()', trade);
    let params = await this.wepay.getBrandWCPayRequestParams(trade);
    if (!params || !(params.package && params.paySign)) {
      throw new Errcode(
        'error! getBrandWCPayRequestParams() return null! orderId=' + orderId,
        EC.ERR_PAY_PARAMS_ERROR
      );
    }

    ctx.body = {
      errcode: 0,
      result: params
    };
  };

  getSandboxSignKey = async (ctx, next) => {
    let params = await this.wepay.getSignKey();
    ctx.body = {
      errcode: 0,
      result: params
    };
  };

  refund = async (ctx, next) => {
    let order = ctx.request.body || {};
    let retobj = await this.wepay.refund({ ...order });
    debug('refund result:', retobj);
    ctx.body = {
      errcode: 0,
      result: retobj
    };
  };
}
